File: packages/hwpx-core/src/package.ts
/**
 * HWPX OPC-style package handling.
 * Ported from Python hwpx/package.py - uses JSZip for async ZIP I/O.
 */

import JSZip from "jszip";
import { parseXml, serializeXml } from "./xml/dom.js";

const _OPF_NS = "http://www.idpf.org/2007/opf/";

const MEDIA_TYPE_EXTENSIONS: Record<string, string> = {
  "image/jpeg": "jpg",
  "image/jpg": "jpg",
  "image/png": "png",
  "image/bmp": "bmp",
  "image/gif": "gif",
  "image/tiff": "tif",
  "image/svg+xml": "svg",
  "image/webp": "webp",
};

function mediaTypeToExtension(mediaType: string): string {
  return MEDIA_TYPE_EXTENSIONS[mediaType.toLowerCase()] ?? "bin";
}

function normalizedManifestValue(element: Element): string {
  const parts = ["id", "href", "media-type", "properties"]
    .map((attr) => (element.getAttribute(attr) ?? "").toLowerCase())
    .filter((v) => v);
  return parts.join(" ");
}

function manifestMatches(element: Element, ...candidates: string[]): boolean {
  const normalized = normalizedManifestValue(element);
  return candidates.some((c) => c && normalized.includes(c));
}

function ensureBytes(value: Uint8Array | string): Uint8Array {
  if (value instanceof Uint8Array) return value;
  return new TextEncoder().encode(value);
}

/** Get the filename portion of a path (like PurePosixPath.name). */
function pathName(path: string): string {
  const idx = path.lastIndexOf("/");
  return idx >= 0 ? path.substring(idx + 1) : path;
}

export class HwpxPackage {
  static readonly MANIFEST_PATH = "Contents/content.hpf";
  static readonly HEADER_PATH = "Contents/header.xml";

  private _parts: Map<string, Uint8Array>;
  private _manifestTree: Document | null = null;
  private _spineCache: string[] | null = null;
  private _sectionPathsCache: string[] | null = null;
  private _headerPathsCache: string[] | null = null;
  private _masterPagePathsCache: string[] | null = null;
  private _historyPathsCache: string[] | null = null;
  private _versionPathCache: string | null = null;
  private _versionPathCacheResolved = false;

  constructor(parts: Map<string, Uint8Array>) {
    this._parts = new Map(parts);
  }

  /** Open an HWPX package from a Uint8Array or ArrayBuffer. */
  static async open(source: Uint8Array | ArrayBuffer): Promise<HwpxPackage> {
    const zip = await JSZip.loadAsync(source);
    const parts = new Map<string, Uint8Array>();
    const promises: Promise<void>[] = [];

    zip.forEach((relativePath, file) => {
      if (!file.dir) {
        promises.push(
          file.async("uint8array").then((data) => {
            parts.set(relativePath, data);
          }),
        );
      }
    });

    await Promise.all(promises);
    return new HwpxPackage(parts);
  }

  // -- Accessors --

  partNames(): string[] {
    return Array.from(this._parts.keys());
  }

  hasPart(partName: string): boolean {
    return this._parts.has(partName);
  }

  getPart(partName: string): Uint8Array {
    const data = this._parts.get(partName);
    if (data == null) {
      throw new Error(`Package does not contain part '${partName}'`);
    }
    return data;
  }

  setPart(partName: string, payload: Uint8Array | string): void {
    this._parts.set(partName, ensureBytes(payload));
    if (partName === HwpxPackage.MANIFEST_PATH) {
      this._manifestTree = null;
      this._spineCache = null;
      this._sectionPathsCache = null;
      this._headerPathsCache = null;
      this._masterPagePathsCache = null;
      this._historyPathsCache = null;
      this._versionPathCache = null;
      this._versionPathCacheResolved = false;
    }
  }

  getXml(partName: string): Element {
    const data = this.getPart(partName);
    const text = new TextDecoder().decode(data);
    const doc = parseXml(text);
    return doc.documentElement;
  }

  setXml(partName: string, element: Element): void {
    const xml = '<?xml version="1.0" encoding="UTF-8"?>' + serializeXml(element);
    this.setPart(partName, xml);
  }

  getText(partName: string): string {
    return new TextDecoder().decode(this.getPart(partName));
  }

  // -- Manifest helpers --

  private manifestTree(): Document {
    if (this._manifestTree == null) {
      const data = this.getPart(HwpxPackage.MANIFEST_PATH);
      const text = new TextDecoder().decode(data);
      this._manifestTree = parseXml(text);
    }
    return this._manifestTree;
  }

  private manifestItems(): Element[] {
    const doc = this.manifestTree();
    const root = doc.documentElement;
    const items: Element[] = [];
    // Walk the DOM tree to find item elements under manifest
    const walk = (node: Element): void => {
      const children = node.childNodes;
      for (let i = 0; i < children.length; i++) {
        const child = children.item(i);
        if (child && child.nodeType === 1) {
          const el = child as Element;
          const tag = el.localName ?? el.tagName;
          if (tag === "item") {
            items.push(el);
          }
          walk(el);
        }
      }
    };
    walk(root);
    return items;
  }

  private resolveSpinePaths(): string[] {
    if (this._spineCache == null) {
      const doc = this.manifestTree();
      const root = doc.documentElement;
      const manifestItems: Record<string, string> = {};

      const findElements = (node: Element, localNameTarget: string): Element[] => {
        const result: Element[] = [];
        const walk = (n: Element): void => {
          const children = n.childNodes;
          for (let i = 0; i < children.length; i++) {
            const child = children.item(i);
            if (child && child.nodeType === 1) {
              const el = child as Element;
              const tag = el.localName ?? el.tagName;
              if (tag === localNameTarget) {
                result.push(el);
              }
              walk(el);
            }
          }
        };
        walk(node);
        return result;
      };

      for (const item of findElements(root, "item")) {
        const id = item.getAttribute("id");
        const href = item.getAttribute("href") ?? "";
        if (id && href) {
          manifestItems[id] = href;
        }
      }

      const spinePaths: string[] = [];
      for (const itemref of findElements(root, "itemref")) {
        const idref = itemref.getAttribute("idref");
        if (!idref) continue;
        const href = manifestItems[idref];
        if (href) {
          spinePaths.push(href);
        }
      }
      this._spineCache = spinePaths;
    }
    return this._spineCache;
  }

  sectionPaths(): string[] {
    if (this._sectionPathsCache == null) {
      let paths = this.resolveSpinePaths().filter(
        (p) => p && pathName(p).startsWith("section"),
      );
      if (paths.length === 0) {
        paths = Array.from(this._parts.keys()).filter((name) =>
          pathName(name).startsWith("section"),
        );
      }
      this._sectionPathsCache = paths;
    }
    return [...this._sectionPathsCache];
  }

  headerPaths(): string[] {
    if (this._headerPathsCache == null) {
      let paths = this.resolveSpinePaths().filter(
        (p) => p && pathName(p).startsWith("header"),
      );
      if (paths.length === 0 && this.hasPart(HwpxPackage.HEADER_PATH)) {
        paths = [HwpxPackage.HEADER_PATH];
      }
      this._headerPathsCache = paths;
    }
    return [...this._headerPathsCache];
  }

  masterPagePaths(): string[] {
    if (this._masterPagePathsCache == null) {
      let paths = this.manifestItems()
        .filter((item) => manifestMatches(item, "masterpage", "master-page"))
        .map((item) => item.getAttribute("href") ?? "")
        .filter((href) => href);

      if (paths.length === 0) {
        paths = Array.from(this._parts.keys()).filter((name) => {
          const n = pathName(name).toLowerCase();
          return n.includes("master") && n.includes("page");
        });
      }
      this._masterPagePathsCache = paths;
    }
    return [...this._masterPagePathsCache];
  }

  historyPaths(): string[] {
    if (this._historyPathsCache == null) {
      let paths = this.manifestItems()
        .filter((item) => manifestMatches(item, "history"))
        .map((item) => item.getAttribute("href") ?? "")
        .filter((href) => href);

      if (paths.length === 0) {
        paths = Array.from(this._parts.keys()).filter((name) =>
          pathName(name).toLowerCase().includes("history"),
        );
      }
      this._historyPathsCache = paths;
    }
    return [...this._historyPathsCache];
  }

  versionPath(): string | null {
    if (!this._versionPathCacheResolved) {
      let path: string | null = null;
      for (const item of this.manifestItems()) {
        if (manifestMatches(item, "version")) {
          const href = (item.getAttribute("href") ?? "").trim();
          if (href) {
            path = href;
            break;
          }
        }
      }
      if (path == null && this.hasPart("version.xml")) {
        path = "version.xml";
      }
      this._versionPathCache = path;
      this._versionPathCacheResolved = true;
    }
    return this._versionPathCache;
  }

  // -- Binary item management --

  /**
   * Add a binary item (image, etc.) to the package.
   * Stores the data in BinData/ and registers it in the manifest.
   * Returns the binaryItemIDRef to use in <hc:img>.
   */
  addBinaryItem(data: Uint8Array, opts: {
    mediaType: string;
    extension?: string;
  }): string {
    // Determine extension from mediaType if not provided
    const ext = opts.extension ?? mediaTypeToExtension(opts.mediaType);

    // Find next available image number
    const existingParts = this.partNames().filter(p => p.startsWith("BinData/"));
    let maxNum = 0;
    for (const p of existingParts) {
      const match = /^BinData\/image(\d+)\./.exec(p);
      if (match?.[1]) {
        const n = parseInt(match[1], 10);
        if (n > maxNum) maxNum = n;
      }
    }
    const nextNum = maxNum + 1;
    const itemId = `image${nextNum}`;
    const href = `BinData/${itemId}.${ext}`;

    // Store binary data
    this._parts.set(href, data);

    // Update manifest: add <opf:item> element
    const manifestDoc = this.manifestTree();
    const root = manifestDoc.documentElement;

    // Find <manifest> element
    let manifestEl: Element | null = null;
    const walk = (node: Element): void => {
      const children = node.childNodes;
      for (let i = 0; i < children.length; i++) {
        const child = children.item(i);
        if (child && child.nodeType === 1) {
          const el = child as Element;
          const tag = el.localName ?? el.tagName;
          if (tag === "manifest") { manifestEl = el; return; }
          walk(el);
        }
      }
    };
    walk(root);

    if (manifestEl) {
      const item = manifestDoc.createElementNS(_OPF_NS, "opf:item");
      item.setAttribute("id", itemId);
      item.setAttribute("href", href);
      item.setAttribute("media-type", opts.mediaType);
      item.setAttribute("isEmbeded", "1");
      (manifestEl as Element).appendChild(item);

      // Persist updated manifest (serializeXml on Document includes <?xml?> declaration)
      const xml = serializeXml(manifestDoc as unknown as Node);
      this._parts.set(HwpxPackage.MANIFEST_PATH, new TextEncoder().encode(xml));
      // Keep cached tree (we just modified it)
    }

    // Invalidate caches that depend on manifest
    this._spineCache = null;
    this._sectionPathsCache = null;
    this._headerPathsCache = null;
    this._masterPagePathsCache = null;
    this._historyPathsCache = null;
    this._versionPathCache = null;
    this._versionPathCacheResolved = false;

    return itemId;
  }

  // -- Saving --

  async save(updates?: Record<string, Uint8Array | string>): Promise<Uint8Array> {
    if (updates) {
      for (const [partName, payload] of Object.entries(updates)) {
        this.setPart(partName, payload);
      }
    }

    const zip = new JSZip();
    for (const [name, data] of this._parts.entries()) {
      zip.file(name, data);
    }
    return zip.generateAsync({ type: "uint8array", compression: "DEFLATE" });
  }
}

File: src/hwpx/opc/package.py
"""Utilities for reading and writing HWPX OPC packages."""

from __future__ import annotations

import io
from dataclasses import dataclass
from pathlib import Path
from typing import BinaryIO, Iterable, Iterator, Mapping, MutableMapping
from zipfile import ZIP_DEFLATED, ZIP_STORED, ZipFile, ZipInfo

from lxml import etree

from .xml_utils import (
    extract_xml_declaration,
    iter_declared_namespaces,
    parse_xml,
    serialize_xml,
)

__all__ = ["HwpxPackage", "HwpxPackageError", "HwpxStructureError", "RootFile", "VersionInfo"]

_OPF_NS = "http://www.idpf.org/2007/opf/"


class HwpxPackageError(Exception):
    """Base error raised for issues related to :class:`HwpxPackage`."""


class HwpxStructureError(HwpxPackageError):
    """Raised when the underlying HWPX package violates the required structure."""


@dataclass(frozen=True)
class RootFile:
    """Represents a ``rootfile`` entry from ``META-INF/container.xml``."""

    full_path: str
    media_type: str | None = None

    def ensure_exists(self, files: Mapping[str, bytes]) -> None:
        """Ensure that the referenced root file actually exists in ``files``."""

        if self.full_path not in files:
            raise HwpxStructureError(
                f"Root content '{self.full_path}' declared in container.xml is missing."
            )


class VersionInfo:
    """Model for the ``version.xml`` document."""

    def __init__(
        self,
        element: etree._Element,
        namespaces: Mapping[str, str],
        xml_declaration: bytes | None,
    ) -> None:
        self._element = element
        self._namespaces = dict(namespaces)
        self._xml_declaration = xml_declaration
        self._dirty = False

    @classmethod
    def from_bytes(cls, data: bytes) -> VersionInfo:
        element = parse_xml(data)
        namespaces = cls._collect_namespaces(data)
        declaration = cls._extract_declaration(data)
        return cls(element, namespaces, declaration)

    @staticmethod
    def _collect_namespaces(data: bytes) -> Mapping[str, str]:
        return iter_declared_namespaces(data)

    @staticmethod
    def _extract_declaration(data: bytes) -> bytes | None:
        return extract_xml_declaration(data)

    @property
    def attributes(self) -> Mapping[str, str]:
        return dict(self._element.attrib)

    def get(self, key: str, default: str | None = None) -> str | None:
        return self._element.attrib.get(key, default)

    def set(self, key: str, value: str) -> None:
        self._element.attrib[key] = value
        self._dirty = True

    @property
    def tag(self) -> str:
        return self._element.tag

    def to_bytes(self) -> bytes:
        xml_body = serialize_xml(self._element, xml_declaration=False)
        if self._xml_declaration:
            return self._xml_declaration + xml_body
        return xml_body

    @property
    def dirty(self) -> bool:
        return self._dirty

    def mark_clean(self) -> None:
        self._dirty = False


class HwpxPackage:
    """Represents an HWPX package backed by an Open Packaging Convention container."""

    CONTAINER_PATH = "META-INF/container.xml"
    VERSION_PATH = "version.xml"
    MIMETYPE_PATH = "mimetype"
    DEFAULT_MIMETYPE = "application/hwp+zip"
    MANIFEST_PATH = "Contents/content.hpf"
    HEADER_PATH = "Contents/header.xml"

    def __init__(
        self,
        files: MutableMapping[str, bytes],
        rootfiles: Iterable[RootFile],
        version_info: VersionInfo,
        mimetype: str,
    ) -> None:
        self._files = files
        self._rootfiles = list(rootfiles)
        self._version = version_info
        self._mimetype = mimetype
        self._manifest_tree: etree._Element | None = None
        self._spine_cache: list[str] | None = None
        self._section_paths_cache: list[str] | None = None
        self._header_paths_cache: list[str] | None = None
        self._master_page_paths_cache: list[str] | None = None
        self._history_paths_cache: list[str] | None = None
        self._version_path_cache: str | None = None
        self._version_path_cache_resolved = False
        self._validate_structure()

    @classmethod
    def open(cls, pkg_file: str | Path | bytes | bytearray | BinaryIO) -> HwpxPackage:
        if isinstance(pkg_file, (bytes, bytearray)):
            stream: str | Path | BinaryIO = io.BytesIO(pkg_file)
        else:
            stream = pkg_file

        with ZipFile(stream, "r") as zf:
            files = {info.filename: zf.read(info.filename) for info in zf.infolist()}
        if cls.MIMETYPE_PATH not in files:
            raise HwpxStructureError("HWPX package is missing the mandatory 'mimetype' file.")
        mimetype = files[cls.MIMETYPE_PATH].decode("utf-8")
        rootfiles = cls._parse_container(files.get(cls.CONTAINER_PATH))
        version_info = cls._parse_version(files.get(cls.VERSION_PATH))
        package = cls(files, rootfiles, version_info, mimetype)
        return package

    @staticmethod
    def _parse_container(data: bytes | None) -> list[RootFile]:
        if data is None:
            raise HwpxStructureError(
                "HWPX package is missing 'META-INF/container.xml'."
            )
        root = parse_xml(data)
        rootfiles = []
        for elem in root.findall(".//{*}rootfile"):
            full_path = (
                elem.get("full-path")
                or elem.get("fullPath")
                or elem.get("full_path")
            )
            if not full_path:
                raise HwpxStructureError("container.xml contains a rootfile without 'full-path'.")
            media_type = (
                elem.get("media-type")
                or elem.get("mediaType")
                or elem.get("media_type")
            )
            rootfiles.append(RootFile(full_path, media_type))
        if not rootfiles:
            raise HwpxStructureError("container.xml does not declare any rootfiles.")
        return rootfiles

    @staticmethod
    def _parse_version(data: bytes | None) -> VersionInfo:
        if data is None:
            raise HwpxStructureError("HWPX package is missing 'version.xml'.")
        return VersionInfo.from_bytes(data)

    def _validate_structure(self) -> None:
        for rootfile in self._rootfiles:
            rootfile.ensure_exists(self._files)
        if not any(path.startswith(("Contents/", "Content/")) for path in self._files):
            raise HwpxStructureError(
                "HWPX package does not contain a 'Contents' directory."
            )

    @property
    def mimetype(self) -> str:
        return self._mimetype

    @property
    def rootfiles(self) -> tuple[RootFile, ...]:
        return tuple(self._rootfiles)

    def iter_rootfiles(self) -> Iterator[RootFile]:
        yield from self._rootfiles

    @property
    def main_content(self) -> RootFile:
        for rootfile in self._rootfiles:
            if rootfile.media_type == "application/hwpml-package+xml":
                return rootfile
        return self._rootfiles[0]

    @property
    def version_info(self) -> VersionInfo:
        return self._version

    def read(self, path: str) -> bytes:
        norm_path = self._normalize_path(path)
        try:
            return self._files[norm_path]
        except KeyError as exc:
            raise HwpxPackageError(f"File '{norm_path}' is not present in the package.") from exc

    def write(self, path: str, data: bytes | str) -> None:
        norm_path = self._normalize_path(path)
        if isinstance(data, str):
            data = data.encode("utf-8")
        pending_rootfiles: list[RootFile] | None = None
        pending_version: VersionInfo | None = None
        if norm_path == self.MIMETYPE_PATH:
            mimetype = data.decode("utf-8")
        elif norm_path == self.CONTAINER_PATH:
            pending_rootfiles = self._parse_container(data)
        elif norm_path == self.VERSION_PATH:
            pending_version = self._parse_version(data)
        self._files[norm_path] = data
        self._invalidate_caches(norm_path)
        if norm_path == self.MIMETYPE_PATH:
            self._mimetype = mimetype
        elif norm_path == self.CONTAINER_PATH:
            assert pending_rootfiles is not None
            self._rootfiles = pending_rootfiles
        elif norm_path == self.VERSION_PATH:
            assert pending_version is not None
            self._version = pending_version
        self._validate_structure()

    def delete(self, path: str) -> None:
        norm_path = self._normalize_path(path)
        if norm_path not in self._files:
            raise HwpxPackageError(f"File '{norm_path}' is not present in the package.")
        if norm_path in {self.MIMETYPE_PATH, self.CONTAINER_PATH, self.VERSION_PATH}:
            raise HwpxStructureError(
                "Cannot remove mandatory files ('mimetype', 'container.xml', 'version.xml')."
            )
        del self._files[norm_path]
        self._validate_structure()

    @staticmethod
    def _normalize_path(path: str) -> str:
        return path.replace("\\", "/")

    def files(self) -> list[str]:
        return sorted(self._files)

    def part_names(self) -> list[str]:
        return self.files()

    def has_part(self, part_name: str) -> bool:
        return self._normalize_path(part_name) in self._files

    def get_part(self, part_name: str) -> bytes:
        return self.read(part_name)

    def set_part(self, part_name: str, payload: bytes | str | etree._Element) -> None:
        if isinstance(payload, etree._Element):
            data = serialize_xml(payload, xml_declaration=True)
        elif isinstance(payload, str):
            data = payload.encode("utf-8")
        elif isinstance(payload, bytes):
            data = payload
        else:
            raise TypeError(f"unsupported part payload type: {type(payload)!r}")
        self.write(part_name, data)

    def get_xml(self, part_name: str) -> etree._Element:
        return parse_xml(self.read(part_name))

    def set_xml(self, part_name: str, element: etree._Element) -> None:
        self.set_part(part_name, element)

    def get_text(self, part_name: str, encoding: str = "utf-8") -> str:
        return self.read(part_name).decode(encoding)

    def manifest_tree(self) -> etree._Element:
        if self._manifest_tree is None:
            self._manifest_tree = self.get_xml(self.MANIFEST_PATH)
        return self._manifest_tree

    def _manifest_items(self) -> list[etree._Element]:
        manifest = self.manifest_tree()
        ns = {"opf": _OPF_NS}
        return list(manifest.findall("./opf:manifest/opf:item", ns))

    @staticmethod
    def _normalized_manifest_value(element: etree._Element) -> str:
        values = [
            element.attrib.get("id", ""),
            element.attrib.get("href", ""),
            element.attrib.get("media-type", ""),
            element.attrib.get("properties", ""),
        ]
        return " ".join(part.lower() for part in values if part)

    @classmethod
    def _manifest_matches(cls, element: etree._Element, *candidates: str) -> bool:
        normalized = cls._normalized_manifest_value(element)
        return any(candidate in normalized for candidate in candidates if candidate)

    def _resolve_spine_paths(self) -> list[str]:
        if self._spine_cache is None:
            manifest = self.manifest_tree()
            ns = {"opf": _OPF_NS}
            manifest_items: dict[str, str] = {}
            for item in manifest.findall("./opf:manifest/opf:item", ns):
                item_id = item.attrib.get("id")
                href = item.attrib.get("href", "")
                if item_id and href:
                    manifest_items[item_id] = href
            spine_paths: list[str] = []
            for itemref in manifest.findall("./opf:spine/opf:itemref", ns):
                idref = itemref.attrib.get("idref")
                if not idref:
                    continue
                href = manifest_items.get(idref)
                if href:
                    spine_paths.append(href)
            self._spine_cache = spine_paths
        return self._spine_cache

    def section_paths(self) -> list[str]:
        if self._section_paths_cache is None:
            from pathlib import PurePosixPath

            paths = [
                path
                for path in self._resolve_spine_paths()
                if path and PurePosixPath(path).name.startswith("section")
            ]
            if not paths:
                paths = [
                    name
                    for name in self._files.keys()
                    if PurePosixPath(name).name.startswith("section")
                ]
            self._section_paths_cache = paths
        return list(self._section_paths_cache)

    def header_paths(self) -> list[str]:
        if self._header_paths_cache is None:
            from pathlib import PurePosixPath

            paths = [
                path
                for path in self._resolve_spine_paths()
                if path and PurePosixPath(path).name.startswith("header")
            ]
            if not paths and self.has_part(self.HEADER_PATH):
                paths = [self.HEADER_PATH]
            self._header_paths_cache = paths
        return list(self._header_paths_cache)

    def master_page_paths(self) -> list[str]:
        if self._master_page_paths_cache is None:
            from pathlib import PurePosixPath

            paths = [
                item.attrib.get("href", "")
                for item in self._manifest_items()
                if self._manifest_matches(item, "masterpage", "master-page")
                and item.attrib.get("href")
            ]
            if not paths:
                paths = [
                    name
                    for name in self._files.keys()
                    if "master" in PurePosixPath(name).name.lower()
                    and "page" in PurePosixPath(name).name.lower()
                ]
            self._master_page_paths_cache = paths
        return list(self._master_page_paths_cache)

    def history_paths(self) -> list[str]:
        if self._history_paths_cache is None:
            from pathlib import PurePosixPath

            paths = [
                item.attrib.get("href", "")
                for item in self._manifest_items()
                if self._manifest_matches(item, "history") and item.attrib.get("href")
            ]
            if not paths:
                paths = [
                    name
                    for name in self._files.keys()
                    if "history" in PurePosixPath(name).name.lower()
                ]
            self._history_paths_cache = paths
        return list(self._history_paths_cache)

    def version_path(self) -> str | None:
        if not self._version_path_cache_resolved:
            path: str | None = None
            for item in self._manifest_items():
                if self._manifest_matches(item, "version"):
                    href = item.attrib.get("href", "").strip()
                    if href:
                        path = href
                        break
            if path is None and self.has_part(self.VERSION_PATH):
                path = self.VERSION_PATH
            self._version_path_cache = path
            self._version_path_cache_resolved = True
        return self._version_path_cache

    def _invalidate_caches(self, changed_path: str) -> None:
        if changed_path == self.MANIFEST_PATH:
            self._manifest_tree = None
            self._spine_cache = None
            self._section_paths_cache = None
            self._header_paths_cache = None
            self._master_page_paths_cache = None
            self._history_paths_cache = None
            self._version_path_cache = None
            self._version_path_cache_resolved = False
        elif changed_path == self.VERSION_PATH:
            self._version_path_cache_resolved = False

    def save(
        self,
        pkg_file: str | Path | BinaryIO | None = None,
        updates: Mapping[str, bytes | str | etree._Element] | None = None,
    ) -> str | Path | BinaryIO | bytes | None:
        if updates:
            for part_name, payload in updates.items():
                self.set_part(part_name, payload)

        destination = pkg_file
        if destination is None:
            buffer = io.BytesIO()
            self._save_to_zip(buffer)
            return buffer.getvalue()

        self._save_to_zip(destination)
        return destination

    def _save_to_zip(self, pkg_file: str | Path | BinaryIO) -> None:
        self._files[self.MIMETYPE_PATH] = self._mimetype.encode("utf-8")
        if self._version.dirty:
            self._files[self.VERSION_PATH] = self._version.to_bytes()
            self._version.mark_clean()
        self._validate_structure()
        with ZipFile(pkg_file, "w") as zf:
            self._write_mimetype(zf)
            for name in sorted(self._files):
                if name == self.MIMETYPE_PATH:
                    continue
                data = self._files[name]
                info = ZipInfo(name)
                info.compress_type = ZIP_DEFLATED
                zf.writestr(info, data)

    def _write_mimetype(self, zf: ZipFile) -> None:
        info = ZipInfo(self.MIMETYPE_PATH)
        info.compress_type = ZIP_STORED
        zf.writestr(info, self._files[self.MIMETYPE_PATH])
